thewalrus/fock_gradients.py

```diff
@@ -0,0 +1,270 @@
+def grad_S2gate(T):
+    """Calculate the gradient of the S2gate
+    Args:
+        T (array): Array representing the gate
+    Returns:
+        (array): Array representing the gradient of the gate
+    """
+    cutoff, _, _, _ = T.shape
+    cutoff -= 1
+    gradT = np.zeros([cutoff, cutoff, cutoff, cutoff])
+    for n in range(cutoff):
+        for k in range(cutoff):
+            for m in range(cutoff):
+                l = m - n + k
+                if 0 <= l < cutoff:
+                    gradT[n, k, m, l] = np.sqrt((m + 1) * (l + 1)) * T[n, k, m + 1, l + 1]
+                    if m > 0 and l > 0:
+                        gradT[n, k, m, l] -= np.sqrt(m * l) * T[n, k, m - 1, l - 1]
+    return gradT
+
+
+def S2gate(s, cutoff, grad=False, r=np.arcsinh(1.0)):
+    r"""
+    Calculates the Fock representation of the S2gate and its gradient
+    Args:
+        s (float): Parameter of the gate
+        cutoff (int): Fock ladder cutoff
+        grad (boolean): Whether to calculate the gradient or not
+        r (float): Value of the parameter used internally in fock_tensor.
+    Returns:
+        tuple(G,dG): The Fock representations of the gate and its gradient
+    """
+    S = two_mode_squeezing(s, 0)
+    if not grad:
+        return fock_tensor(S, np.zeros([2]), cutoff, r=r).real, None
+
+    T = fock_tensor(S, np.zeros([2]), cutoff + 1, r=r).real
+    return T[0:cutoff, 0:cutoff, 0:cutoff, 0:cutoff], grad_S2gate(T)
+
+
+@jit("double[:,:,:,:](double[:,:,:,:])")
+def grad_BSgate(T):
+    """Calculates the gradient of the BSgate
+    Args:
+        T (array): Array representing the gate
+    Returns:
+        (array): Array representing the gradient of the gate
+    """
+    cutoff, _, _, _ = T.shape
+    cutoff -= 1
+    gradT = np.zeros([cutoff, cutoff, cutoff, cutoff])
+    for n in range(cutoff):
+        for k in range(cutoff):
+            for m in range(cutoff):
+                l = n + k - m
+                if 0 <= l < cutoff:
+                    if m > 0:
+                        gradT[n, k, m, l] = np.sqrt(m * (l + 1)) * T[n, k, m - 1, l + 1]
+                    if l > 0:
+                        gradT[n, k, m, l] -= np.sqrt((m + 1) * l) * T[n, k, m + 1, l - 1]
+    return gradT
+
+
+def BSgate(theta, cutoff, grad=False, r=np.arcsinh(1.0)):
+    r"""
+    Calculates the Fock representation of the BSgate and its gradient
+    Arg:
+        theta (float): Parameter of the gate
+        cutoff (int): Fock ladder cutoff
+        grad (boolean): Whether to calculate the gradient or not
+        r (float): Value of the parameter used internally in fock_tensor.
+    Returns:
+        tuple(G,dG): The Fock representations of the gate and its gradient
+    """
+    S = beam_splitter(theta, 0)
+    if not grad:
+        return fock_tensor(S, np.zeros([2]), cutoff, r=r).real, None
+
+    T = fock_tensor(S, np.zeros([2]), cutoff + 1, r=r).real
+    return T[0:cutoff, 0:cutoff, 0:cutoff, 0:cutoff], grad_BSgate(T)
```

Comment:
Will a Dgate function be added once it is determined the best way to compose it?